### ISSUE TITLE
Unfinished spans and multiple-finish() undefined behaviour

### ISSUE THREAD

**Issue by lookfwd** [opentracing/opentracing.github.io#85]
Hello, I have a parent `Span`. It has several children on many servers. Some of those child-`Span`s are on processes that have e.g. 2-minute timeouts. On timeout further children `Span`s on other servers might be injected/joined. I have a trace id and I want to investigate what's going on 1 minute after the parent `Span` started.

What's the conceptual state of the tracing system?

Should I not be able to see on my tools the parent `Span` at all because it didn't `.finish()` yet? In that case what happens if it never `.finish()`'s? I should not have a trace for the single most interesting case of traces?

Should I see the state of child `Span`s (which have `.finish()`'ed) without having any information for the parent `Span` (apart from its ID and baggage)?

Should I see all the information I have available for the parent `Span`? How do I provide useful information in this case since I can't call `.finish()` (undef behaviour on double-`finish()`? Is there the need for `.flush()`? Are all `Span`s then potentially infinite and we present just - to the best of our knowledge - information up to the moment?

This isn't a new feature. Existing systems handle those cases in some way. The extend to which they see `Span`s as transactions or a group of streaming events varies. Opentracing must specify this behaviour. There are children that outlive parents, bugs, system crashes, network delays, all sorts of asynchronous behaviour and this is specifically why we need tracing. This isn't a corner case. This is the use case.

**Comment by dkuebric** [comment]
@lookfwd this is a tricky problem in the general case.  For example, in the event of a crash in the traced process, any in-flight spans will definitely not be finished. 

In AppNeta's X-Trace implementation, we actually use an event-based model where a span is composed of two events.  So we will have received information available at the start of a span even if the end of the span never occurs for some reason.  However, there is still the possibility for missing exit events.

So that's one way to mitigate the risk.  Additionally, it's possible that one could put some internal span timeouts to ensure span delivery, albeit with incomplete data and/or missing end time.

I think that the tracing backend would ultimately have to take some responsibility for this due to the possibility of entirely missing spans, however.  The experience on the backend is a missing span, which is referenced by any downstream spans (via parent ids).

**Comment by lookfwd** [comment]
@dkuebric perfectly agree! I find this tricky problem very interesting myself. This is where a specification should have mechanisms in place to ensure implementations have well-thought behaviour.

> In AppNeta's X-Trace implementation, we actually use an event-based model where a span is composed of two events.

Nice! Yes, I see how this is a nice pragmatic approach.

> However, there is still the possibility for missing exit events.

That's not a problem at all. Problem is when you specify "Unless everything is done perfectly, the standard guarantees you nothing". It sets the bar too low.

> any in-flight spans will definitely not be finished

Yes, exactly! Infinite spans is fact of life... The question is how tracing systems should deal with them.

> internal span timeouts

or/and an explicit `.flush()`. If I'm a parent span and I spawn 10 async calls, database writes or whatever, I've mostly done my job (potentially in μs) and I know it... no need for (potentially long) timeouts. I can `.flush()` and on the event of a crash I've done my best. It's up to the implementation what durability guarantees it gives at that point. When all child spans finish (maybe minutes later) I (parent span) can also call my `.finish()` and everything is even more happy and perfect.

Another possibility is that child Spans might `ack` the transaction (in memory) and at that point the parent span is ok to `.finish()`. Unfortunately this doesn't mean that child `Span`s `.finish()`ed. Child `Span`s ` might `finish()` when the transaction is e.g. on the commit log, on disk. This is a very typical async pattern [e.g.](https://wiki.apache.org/cassandra/Durability).

> I think that the tracing backend would ultimately have to take some responsibility for this due to the possibility of entirely missing spans, however.

Absolutely! The backend should do the best possible, given the information it has, to provide it to the user. As a backend I shouldn't e.g. drop a trace because some info seems incomplete.

**Comment by bg451** [comment]
@lookfwd re: `flush()` We had an issue a while ago about adding `flush()`. 

https://github.com/opentracing/opentracing.github.io/issues/74

**Comment by lookfwd** [comment]
I don't see anything contrary in #74 in the context `flush()` is discussed here.

> There doesn't seem to be a strong production use case for this

I think I gave enough "production use cases" above.

> the overall semantic meaning of flush() is not clear

Just in case clarifying this is semantically `span.flush()` and its aim is to force implementors to think about taking appropriate actions to ensure durability of the trace up to this point in time since, beyond it, it might - for long time - not hear any updates about it (and a crash might happen in the meanwhile).

Implementations could do a `nop()` but they will have to state it on their documentation (so that people who need durability guarantees are aware of the implementation not fitting their needs).

> we would need to debate which exact async API

`Span` should have enough knowledge to either implement or ignore this method call. If a span knows how to `finish()` it should know how to `flush()`.

**Comment by lookfwd** [comment]
Here's a diagram of some of the things I mention above.

![use-case](https://cloud.githubusercontent.com/assets/789359/14359485/8829885c-fce9-11e5-958c-ca9469842159.PNG)

**Comment by yurishkuro** [comment]
This use case seems to arise only with async children. We had discussions in the past how OT API can provide the means of expressing the fact that a child span is started to represent async work and is not on the critical path of the parent. I can't find an open issue for this at the moment, but I think if we address that, it will implicitly address the same problem as what the `span.flush()` is trying to address, only in a more direct way.

**Comment by yurishkuro** [comment]
iirc, I believe the approach we were leaning towards was by setting some special "Async" tag on the child span. That should provide enough signal to the tracing system that it might want to do something special with the parent (like "flush").

**Comment by lookfwd** [comment]
Async is one use case. Existence of timeouts is another in which case you need to flush early.

**Comment by lookfwd** [comment]
I'm wondering - how does tagging something on the child, which might be started with many different ways (e.g. called both synchronously or asynchronously from different services), is easier than `parent.flush()` from an implementation's perspective.

**Comment by jmacd** [comment]
I don't see why a flush API is needed. Can't an implementation keep track of its active Spans and internally decide to flush long-lived potentially infinite spans?

The "multiple-finish" subject is related, but really there are two semantics operations:
- mark by completion time for latency analysis
- note that I will never use this object again
Currently, finish means both of those at once. Neither of these are necessary for an implementation to flush span details. I think it's a possibly a mistake to require that Span object lifetime be tied to span latency. I would recommend that finish() mark the completion of the operation but not end the lifetime of the span object. If finish() is called twice, the first one wins and the subsequent call can be logged as an event on the span. This requires the implementation to support logging span detail after finish.

Yuri raises a related but independent point about "async" tags to indicate that a span is off the critical path. This seems important regardless of how you handle the multiple-finish problem.

**Comment by lookfwd** [comment]
> Can't an implementation keep track of its active Spans and internally decide to flush long-lived potentially infinite spans?

Sure! `.flush()` is application's hint to the implementation... and a reminder to the implementor that a) infinite spans exists and that b) it's _very very good for you_ if you don't wait for `.finish()` to submit your data (because by the way - it might never happen). We need considerate behavior to be the norm, not the exception. If flush is a `nop()` it should be documented.

> there are two semantics operations

Fantastic observation!

> mark by completion time for latency analysis

Excellent - this is perfectly fine! You might even, potentially want to have many such "checkpoints". I might want to mark both the "database write finished checkpoint" and the "upload video to S3 finished checkpoint".

> note that I will never use this object again

This is perfectly unnatural as you mention. It might never happen or it might happen multiple times (due to - among others - a bug) and the tracing system shouldn't freak out if this is the case.

> the first one wins

"the first one wins" on the latency measure... ok. Although checkpointing would be more clear.

> subsequent call can be logged as an event on the span

Yes - this is the most important thing!

> is off the critical path

I understand... in general I get the feeling that most people here care about some "UI latency" and they want it e.g. to be below 50ms. This is why there's such emphasis on the "critical path". In our applications we care a lot about reliability and we have several such "critical paths" since different API consumers have different latency expectations. Some are APIs, some other are UIs and some are just how much risk we feel can handle related to durability. There isn't a single critical path. We also have many replicas and a write goes to all of them. From customer's perspective, one write is enough. From durability's perspective, we need at least quorum across datacenters. From overall health of the cluster and monitoring, writes should get acked from every node for us to be happy (although this doesn't always happen and it's ok).

As I've send in another issue - for me very important is what actually happens on the datacenter. If there are any funny or simplifying assumptions I prefer to have them in the UI than on the collector. I can't afford to go re-annotate all the services every time opentracing changes. I'm ok to evolve both database schemas and UIs to cover more use cases. But annotation should - if possible - happen only once.

**Comment by yurishkuro** [comment]
@lookfwd

> I'm wondering - how does tagging something on the child, which might be started with many different ways (e.g. called both synchronously or asynchronously in different contexts), is easier than parent.flush() from implementation's perspective?

What I mean is that I would rather have instrumentation code describe the actual business scenario, not tell the tracer what to do. Given the vendor-neutral nature of the OT API, a `flush()` method as a hint to the tracer is extremely vague. A tracer can always do a better job if it's given higher fidelity signals, such as those that describe the actual causality relationship between spans. The value of `flush()` method as a causality signal is zero.

**Comment by bensigelman** [comment]
@lookfwd given the (a) number and (b) complexity of the various issues you've brought up here and in the C++ PR, would it be possible to schedule a google hangout or similar to communicate with higher bandwidth? I am happy to coordinate and find a time that works for any parties who'd like to be involved. (cc @dkuebric @bg451 @yurishkuro @jmacd)

A short summary of my position on this: it would be great if OT had standardized ways (tags? logEvent payloads? that sort of thing) of describing arbitrary causal relationships. This would allow OT to represent X-Trace style causality DAGs... even things that don't fit the span model well.

Finally, a clarification: @lookfwd, I think you mean "unfinished span" more than "infinite span", right? "Infinite span" is an oxymoron (since the span concept involves a, well, time *span*). But an "unfinished span" is merely a span whose finish time has not yet been determined. There is nothing in OT that prevents an implementation from recording incomplete Spans incrementally, and indeed I think that would be a good design decision when attempting to handle the case you describe at the top of this issue.

**Comment by lookfwd** [comment]
> I think you mean "unfinished span" more than "infinite span", right? 

Absolutely! "un-finished" / "in-finite" or even "open span" in the sense of "open to new information" and as of half-open interval `[t0, now)` (<- I get philosophical here). Exactly as you say... having a definite start but not necessarily known finish.

> There is nothing in OT that prevents an implementation

There is nothing that enforces it either though. You can't specify a car leaving breaks as implementation specific. This discussion is about the breaks of the car. To be always ready for a crash, any valid implementation should stream as early as possible to the collector. I'm ok with this being just documented as a requirement, but it's even better if there's a mechanism (e.g. `.flush()` hint) to make every implementor to a) think about it and b) document exactly the guarantees they provide. Any such mechanism would do - `.flush()` might be a bad idea.

> a span whose finish time has not yet been determined

This is little detail - but again, I can see how an implementor relying too much on the idea of a (single) finish time adds unnecessary constraints along the way. Here's a case I described previously:

<img width="740" alt="screen shot 2016-04-08 at 04 51 49" src="https://cloud.githubusercontent.com/assets/789359/14373786/a7d030dc-fd45-11e5-86a7-e36d0b9e0951.png">

For those complex cases and beyond you really need tracing because generic monitoring infrastructure isn't sufficient.

The way to get complex cases traced right, is by providing primitives i.e. `Span`s, a way to make them accessible and a way to track causality and let people trace whatever actually happens. Query-side developers can then aggregate and present information in whatever way is meaningful for their target audience.

As @jmacd says, `.finish()` is overloaded - it's both an event and a "I will never use this object again" statement. I don't see why as an event it's particularly special. On the image above there are 2 important events (finish times - t0, t1) which might be subject to SLAs. As an "I will never use this object again" statement, it's not opentracing's business. opentracing might or might not receive a `.finish()` for multiple reasons and it should be functional nevertheless.

> would it be possible to schedule a google hangout or similar to communicate with higher bandwidth?

Absolutely, whatever is convenient.

**Comment by bensigelman** [comment]
Thanks for all of the detail.

> > would it be possible to schedule a google hangout or similar to communicate with higher bandwidth?
> Absolutely, whenever convenient.

Cool, let's schedule a time to talk more interactively, then... Who else would like to participate? (No pressure, of course!)

**Comment by jmacd** [comment]
I like where this is heading. You could have inter-trace links using a set of conventional relationships that we would standardize, like:

blocked_by (the linked span context blocked this span on a shared resource)
blocker_of (the linked span context was blocked by this span on a shared resource)
deferred_child (the linked span context was caused by this span, but the association was severed)
etc.

(I'll be happy to participate in a discussion.)

**Comment by yurishkuro** [comment]
@bensigelman count me in for a conf call. Maybe @lookfwd can make the next week's regular call, Apr 13, 7:30 EST, but other time is fine too.

**Comment by bensigelman** [comment]
@lookfwd IIRC you are in England??? What time ranges work for you (generally speaking, anyway)?

**Comment by bg451** [comment]
Count me in as well!

**Comment by lookfwd** [comment]
Great, 19:30 est this Wednesday. I will try to join. Yes, I'm in London, UK. But I'm a night guy... :)

**Comment by bensigelman** [comment]
(sorry about that! hit the wrong button... ^^^)

**Comment by lookfwd** [comment]
What would you think about something like [this](https://github.com/opentracing/opentracing.github.io/pull/89)? I tried it to be gentle but still put the concept of "unfinished spans" on the map and make implementers think and comment on it.
